libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -1056,36 +1018,13 @@ void AC_PosControl::run_xy_controller(float dt)
         _vehicle_horiz_vel.x = _inav.get_velocity().x;
         _vehicle_horiz_vel.y = _inav.get_velocity().y;
     }
-
-    // calculate velocity error
-    _vel_error.x = _vel_target.x - _vehicle_horiz_vel.x;
-    _vel_error.y = _vel_target.y - _vehicle_horiz_vel.y;
-    // TODO: constrain velocity error and velocity target
-
-    // call pi controller
-    _pid_vel_xy.set_input(_vel_error);
-
-    // get p
-    vel_xy_p = _pid_vel_xy.get_p();
-
-    // update i term if we have not hit the accel or throttle limits OR the i term will reduce
-    // TODO: move limit handling into the PI and PID controller
-    if (!_limit.accel_xy && !_motors.limit.throttle_upper) {
-        vel_xy_i = _pid_vel_xy.get_i();
-    } else {
-        vel_xy_i = _pid_vel_xy.get_i_shrink();
-    }
-
-    // get d
-    vel_xy_d = _pid_vel_xy.get_d();
-
+    Vector2f accel_target = _pid_vel_xy.update_all(_vel_target.x, _vel_target.y, _vehicle_horiz_vel);
```

Comment:
seems a little strange to be passing the x y components of that target separately by then passing the true vel as a vector. Maybe a existing issue of course.

Reply:
I agree. We have some development left before this will finish evolving to its final form. I have not worked out how to best handle saturation and limits yet. Some solution require passing the values of the variable back and overriding the target.  Given that this is a Vector 2f operation on a Vector 3f variable and a 2.5D controller I am not sure how to best handle this. I believe that the best way forward will become clear after all the elements of the position controller are in place. At the moment we are still only half way through.

Reply:
With the code as it stands we can use the Vector2f and Vector3f input so I have changed it as suggested.